Don't treat render targets like textures.

vanruesc opened the pull request:
Hi,

in #8615 I outlined some difficulties that arised during an attempt to move away from treating render targets like textures. @tschw informed me about relevant changes that paved the way for a straight-forward solution to all this.

In #8417 I stated that I'd start with refactoring the examples, but it turned out that this doesn't work anymore because of [this line](https://github.com/mrdoob/three.js/pull/8606/commits/3c354341e14d6ba8e04b6ee2698a6b613453bd70#diff-c0e88b98497597a015ecf238e91ac3a0R3042). If you use the `texture` property of a `CubeCamera`'s `renderTarget` as a map right now, it will break things. So I chose to do it the other way around and started with refactoring the library.

Warnings regarding the use of render targets as textures won't be spammed. I used IIFEs for better encapsulation for the backwards compatibility code and marked the code that might not be easily identified as back-compat code.

In an earlier [diff comment](https://github.com/mrdoob/three.js/pull/8417/files#r57252265) @mrdoob suggested to use `material.envMap instanceof THREE.CubeTexture` instead of a dynamic flag and I replied that the CubeTexture wasn't involved in the respective logic. However, after thinking about alternatives for a while I eventually realised that I was wrong. It's possible to just use the fact that the CubeTexture is not involved directly, because it's involved indirectly. I'm not too familiar with the codebase and that got in my way of thinking.

``` js
uniforms.flipEnvMap.value = ( material.envMap instanceof THREE.WebGLRenderTargetCube ) ? 1 : - 1;
```

can be replicated as

``` js
uniforms.flipEnvMap.value = ( ! ( material.envMap instanceof THREE.CubeTexture ) ) ? 1 : - 1;
```

With this, `WebGLRenderTargetCube` and `Texture` can be left untouched.

I made sure that all the examples work with these changes. I also refactored and checked them again for a seperate PR in case this one gets merged.

vanruesc committed:
Don't treat render targets like textures.

Expect Texture instances from now on. Added backwards compatibility
measures to ensure that render targets can still act as textures for the
time being. A deprecation warning will be given. Resolves #8615.

tschw commented on src/renderers/WebGLRenderer.js:
```diff
@@ -3017,47 +3021,96 @@ THREE.WebGLRenderer = function ( parameters ) {
+
+			} else {
+
+				// assumed: texture property of THREE.WebGLRenderTargetCube
```
How would we get here in the future? Can't see this case being handled.

vanruesc replied:
The `WebGLRenderTargetCube` uses a simple `Texture` instance. That's what I rely on here. So in the future, `setTextureCube` will receive the texture instance of the cube render target which is not a `CubeTexture`.

tschw replied:
And, yes, it's a bad idea to rely on quirky stuff like this - especially without commenting on it.

tschw replied:
I'm talking about a comment in the source code. One on the ticket won't help whoever edits next.

vanruesc replied:
I'm just trying not to change too much at once here. I was hoping that `// assumed: texture property of THREE.WebGLRenderTargetCube` was enough here, sorry if that didn't convey the whole idea.

tschw replied:
It **was**. But now that's no longer what's assumed, is it? :)

tschw replied:
...because `texture.texture` is peeled before that.

It was me writing that comment and I didn't even know about the quirkiness until you actually documented it. Knowing stuff like this is valuable when editing the code and shouldn't be left to rot on some closed ticket, just so it has to be discovered once again by someone else.

vanruesc replied:
No it was `// assumed: texture instanceof THREE.WebGLRenderTargetCube` before and it's now assumed to be a property of a cube render target :(

tschw commented on src/renderers/WebGLRenderer.js:
```diff
@@ -3017,47 +3021,96 @@ THREE.WebGLRenderer = function ( parameters ) {
+
+				setTextureCube( texture, slot );
+
+			} else {
```
I guess we should make it `else if ( texture instanceof THREE.WebGLRenderTargetCube )` and move the warning here.

vanruesc replied:
Certainly possible, but it would result in the same logic. I moved the check and the warning up to make it easier to remove it later.

WestLangley commented on src/renderers/WebGLRenderer.js:
```diff
@@ -2901,7 +2905,7 @@ THREE.WebGLRenderer = function ( parameters ) {
 	}
 
-	function setCubeTexture ( texture, slot ) {
+	function setTextureCube ( texture, slot ) {
```
Function renamed for a reason, or is it just personal preference?

vanruesc replied:
I renamed these for consistency reasons. I didn't see why they were named differently.

tschw replied:
I left the names different because both do different things. Local `setCubeTexture` did not handle the render target case, while public `setTextureCube` did (and still does).

vanruesc replied:
I see. Well, the function names didn't really reflect that difference and there is no documentation about this. But I understand that and will revert these changes.

tschw replied:
Please see my main comment, before doing rushed edits.

vanruesc replied:
Alright, the whole thing just appeared a bit imperative to me.

tschw replied:
Aww, sorry! Actually, just meant to fill in the history why the code was like you found it.

> Well, the function names didn't really reflect that difference and there is no documentation about this.

And this is true too :).

tschw commented on src/renderers/WebGLRenderer.js:
```diff
@@ -2806,8 +2812,6 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 	function setTexture2D( texture, slot ) {
 
-		if ( texture instanceof THREE.WebGLRenderTarget ) texture = texture.texture;
```
Where did this check (2D case) go? Also should warn, I guess?

tschw replied:
Also called internally. Did you check it's safe in all cases?

vanruesc replied:
the local `setTexture2D` function is only called in [this](https://github.com/mrdoob/three.js/pull/8658/files/4c4a9501a88f1e4d08fbd59306faeb18c769e890#diff-c0e88b98497597a015ecf238e91ac3a0R3052) line. The check went just above that call.

tschw replied:
Ah good! That one looks just fine. 

It'd be ideal to use the same scheme for the cube case.

tschw commented:
Wouldn't it be cleaner/clearer to just let `RenderTargetCube` use a `CubeTexture` as I suggested on your ticket?

Right now, it seems rather difficult to get the removal of the deprecation right without leftover. This (*) case.

I think the local function `setTextureCube` should be cleaned up (I think it should be quite easy to express that code in a less scary way), always be called, and `setTextureCubeDynamic` should be removed.

tschw commented on src/renderers/WebGLRenderer.js:
```diff
@@ -3017,47 +3021,96 @@ THREE.WebGLRenderer = function ( parameters ) {
+			}
+
+			if ( texture instanceof THREE.CubeTexture ||
+				 ( Array.isArray( texture.image ) && texture.image.length === 6 ) ) {
```
(*)

vanruesc commented:
First of all, thank you very much for the quick feedback. I didn't expect you to immediately go at it like that :)

> Wouldn't it be cleaner/clearer to just let RenderTargetCube use a CubeTexture as I suggested on your ticket?

Maybe, yes. But I think there must be a reason why `WebGLRenderTargetCube` uses a simple Texture. I'd have to investigate this if I were to go that route, but unfortunately I don't have much more time to spare.

> Right now, it seems rather difficult to get the removal of the deprecation right without leftover.

I actually tried to make it obvious by introducing the iifes. I guess that backfired.

```js
this.setTexture2D = ( function() {

	var warned = false;

	return function( texture, slot ) {

		if ( texture instanceof THREE.WebGLRenderTarget ) {

			if ( ! warned ) {

				console.warn( "THREE.WebGLRenderer.setTexture2D: ..." );
				warned = true;

			}

			texture = texture.texture;

		}

		setTexture2D( texture, slot );

	};

}() );

// would turn into 

this.setTexture2D = setTexture2D;
```

```js
this.setTextureCube = ( function() {

	var warned = false;

	return function( texture, slot ) {

		if ( texture instanceof THREE.WebGLRenderTargetCube ) {

			if ( ! warned ) {

				console.warn( "THREE.WebGLRenderer.setTextureCube: ..." );
				warned = true;

			}

			texture = texture.texture;

		}

		if ( texture instanceof THREE.CubeTexture ||
			 ( Array.isArray( texture.image ) && texture.image.length === 6 ) ) {

			// CompressedTexture can have Array in image :/

			setCubeTexture( texture, slot );

		} else {

			// assumed: texture property of THREE.WebGLRenderTargetCube

			setCubeTextureDynamic( texture, slot );

		}

	};

}() );

// would turn into

this.setTextureCube = function( texture, slot ) {

	if ( texture instanceof THREE.CubeTexture ||
		 ( Array.isArray( texture.image ) && texture.image.length === 6 ) ) {

		// CompressedTexture can have Array in image :/

		setCubeTexture( texture, slot );

	} else {

		// assumed: texture property of THREE.WebGLRenderTargetCube

		setCubeTextureDynamic( texture, slot );

	}

};

```

My aim was to get the render targets out of the picture entirely because their only purpose here is to peel `texture.texture`.

The check whether a uniform is a `WebGLRenderTarget`(`Cube`) is performed in the `WebGLUniforms` class which also later determines if `setTexture2D` or `setTextureCube` is called, if I'm not mistaken.

> I think the local function setTextureCube should be cleaned up (it should be quite easy to express that code in a less scary way), always be called, and setTextureCubeDynamic should be removed.

I agree on that, but I don't have enough insight on that and I think it could be done afterwards.

tschw commented:
> I think there must be a reason why WebGLRenderTargetCube uses a simple Texture.

Yep. The reason is mipmapping. It would have to become optional.

> The check whether a uniform is a WebGLRenderTarget(Cube) is performed in the WebGLUniforms class which also later determines if setTexture2D or setTextureCube is called, if I'm not mistaken.

`WebGLUniforms` only knows whether it's 2D or Cube, not whether the user passes a render target instead of a texture, because it gets the info from the GLSL. It then calls `renderer.setTexture2D` or `.setTextureCube`. But I guess you meant that :)

> I actually tried to make it obvious by introducing the iifes. I guess that backfired.

OK, I get it. And I like it. Please leave it that way. 

However, it invisibly changed the detection. So you should really leave a hint in the source at `CubeRenderTarget` actually having a `.texture` property being a plain `Texture` and **not** a `CubeTexture`! That's something not at all easy to expect.

> I don't have much more time to spare.

:( 

Hope you don't only say so because my review felt too demotivating. It's always about what does not (yet) fit and intermixed with suggestions and seemingly stupid questions, so it can feel as if your submission is perceived in a much more negative way than actually the case.

> > setTextureCube should be cleaned up

> I don't have enough insight on that and I think it could be done afterwards.

Fair enough. Add a comment and let's get it merged then.

vanruesc commented:
> Hope you don't only say so because my review felt too demotivating.

Oh no, don't worry. If I was upset, I'd just tell you straight up :D I'm in the middle of my master thesis right now which is why I really shouldn't dawdle around too much.

> WebGLUniforms only knows whether it's 2D or Cube, not whether the user passes a render target instead of a texture

And the goal is to ban WebGLRenderTargets from being uniforms. I believe that checking for render targets in `setTexture2D` and `setTextureCube` is appropriate, but I must admit that I wasn't too sure about that anymore after your constructive remarks!

> However, it invisibly changed the detection. So you should really leave a hint in the source at CubeRenderTarget actually having a .texture property being a plain Texture and not a CubeTexture! That's something not at all easy to expect.

You're definitely right, will do this first thing in the morning!

vanruesc commented:
Sorry, it took me a while because I was trying to figure out how the identification of render target(cube) textures is being handled. I couldn't find the code that is responsible for classifying them, but here's what I think happens:

- if there's a custom uniform marked as 't', it can only be a samplerCube if it's truly a `CubeTexture` 
- if some object is set as `material.envMap` it's always considered to be some sort of cube texture

Is that correct? I might be completely wrong.

Going from there, I assume that the distinction between calling setTexture2D and setTextureCube is reliable as it is right now and the peeling of `texture.texture` doesn't affect this preliminary decision.

Now if `WebGLRenderTargetCube` starts using `CubeTexture` in the future and the `setTextureCube` function is cleaned up, then that `CubeTexture` instance of the render target is treated just like any other `CubeTexture` (loaded with the `CubeTextureLoader` for example), right?

I'm having a hard time wrapping my head around this. I really didn't expect this to be so confusing.

Anyway, I'm adding the requested comments now to hightlight the quirky parts.

One more thing: I think that it's necessary to add `instanceof THREE.WebGLRenderTarget` to the `setTextureCube` peeling step because there are actually a few examples that use `WebGLRenderTarget` as envMaps (PMREM). This case was implicitly handled before which is why I overlooked it. Should I add it?

tschw commented:
> if there's a custom uniform marked as `'t'`...

Can't be - `'t'` is gone. So is `.type`. I removed it. The info whether a sampler expects a 2D or Cube texture comes from `WebGLActiveInfo` (a WebGL structure - not a `THREE` class) that describes a uniform in the GLSL program. Also it's an error to bind a `sampler2D` to a cube texture or other way around.

> Now if WebGLRenderTargetCube starts using CubeTexture in the future and the setTextureCube function is cleaned up, then that CubeTexture instance of the render target is treated just like any other CubeTexture (loaded with the CubeTextureLoader for example), right?

This is what I had in mind, yes.

Mipmapping will have to become optional (on by default for textures, off by default for render targets) for that, unless of course it is switchable already.

For the loader part, I guess you wouldn't want to save it in the first place, as it's a "target", after all :).

> I'm having a hard time wrapping my head around this. I really didn't expect this to be so confusing.

There's always one or the other confusing surprise with grown code once start digging...

> Anyway, I'm adding the requested comments now to hightlight the quirky parts.

:+1: 

> One more thing: I think that it's necessary to add instanceof THREE.WebGLRenderTarget to the setTextureCube peeling step because there are actually a few examples that use WebGLRenderTarget as envMaps (PMREM). This case was implicitly handled before which is why I overlooked it. Should I add it?

Could you confirm this with a debugger or is it more of a guess?

I think we're actually talking software cubemapping here. For PBR, you need seamless cubes and two dimensions of texture LOD (one for pixel-level filtered mipmaps and the other one for the spatially filtered roughness convolution). A cube sampler won't suffice in this case.

If what you say is actually happening (and there's a cube sampler involved), the examples are the part that should be changed - not the renderer.

vanruesc commented:
> Can't be - 't' is gone. So is .type. I removed it.

Ohh, I see. I haven't quite caught up to that yet. This explains things.

I don't know anything about PBR (yet) so I can't quite follow you, but in the following lines of code from the examples you can see that a `WebGLRenderTarget` is bound to `material.envMap`:

[materials envmap hdr](https://github.com/mrdoob/three.js/blob/dev/examples/webgl_materials_envmaps_hdr.html#L263)
[materials standard](https://github.com/mrdoob/three.js/blob/dev/examples/webgl_materials_standard.html#L203)
[pmremCubeUVPacker.CubeUVRenderTarget](https://github.com/mrdoob/three.js/blob/dev/examples/js/pmrem/PMREMCubeUVPacker.js#L34)

That's what led me to believe that the additional check in `setTextueCube` is necessary. Without it, the texture won't be peeled.

vanruesc committed:
Added comments to clear up RT-Cube detection.

tschw commented:
@tschw
> software cubemapping

I meant it's probably a single 2D texture encoding a cube map.

Just set a breakpoint in the renderer to know for sure whether the code gets called or not.

tschw commented on src/renderers/WebGLRenderer.js:
```diff
@@ -2065,7 +2065,10 @@ THREE.WebGLRenderer = function ( parameters ) {
+		// don't flip CubeTexture envMaps, flip everything else:
+		//  WebGLRenderTargetCube will be flipped for backwards compatibility
+		//  WebGLRenderTargetCube.texture will be flipped because it's a Texture and NOT a CubeTexture
+		// this check must be handled differently, or removed entirely, if WebGLRenderTargetCube uses a CubeTexture in the future
```
This is good! I wouldn't have thought about documenting it here as well. :+1:

tschw commented on src/renderers/WebGLRenderer.js:
```diff
@@ -3093,11 +3101,13 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 			}
 
+			// currently relying on the fact that WebGLRenderTargetCube.texture is a Texture and NOT a CubeTexture
```
:+1:

tschw commented on src/renderers/WebGLRenderer.js:
```diff
@@ -3074,12 +3080,14 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 	}() );
 
+	//this.setTextureCube = setTextureCube;
```
Wouldn't add this one yet.

tschw commented on src/renderers/WebGLRenderer.js:
```diff
@@ -3049,6 +3054,7 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 			}
 
+			// in the future, texture will be expected to be a Texture instance
```
Unnecessary. And I wouldn't add comments starting with "in the future". Noone will care LOL :).

vanruesc replied:
True. Wasn't sure about this.

tschw commented on src/renderers/WebGLRenderer.js:
```diff
@@ -3093,11 +3101,13 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 				// CompressedTexture can have Array in image :/
 
+				// this function alone should take care of cube textures in the future
```
Also not ending with "in the future". I'd say
```javascript
// TODO: Unify these code paths
```
above the `if`...

vanruesc replied:
Got it.

vanruesc commented on src/renderers/WebGLRenderer.js:
```diff
@@ -3030,10 +3033,12 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 	this.allocTextureUnit = allocTextureUnit;
 
+	//this.setTexture2D = setTexture2D;
```
Should I also remove this one?

tschw replied:
No! Please keep it!

Why? These routines are actually interchangeable minus deprecation, the others are not.

vanruesc replied:
Alright, thanks for the help!

vanruesc commented:
> Just set a breakpoint in the renderer to know for sure whether the code gets called or not.

I observed `setTextureCube` in the debugger and the `WebGLRenderTarget` did NOT appear there. It did get passed into `setTexture2D`. 
My assumption stemmed from the idea that envMaps would imply the use of `setTextureCube`, which is wrong. So yeah, false alarm!

However, if I modify the example so that the `texture` property of `CubeUVRenderTarget` is used directly, the following warning message appears and the envMap is not rendered:

```
WebGL: INVALID_OPERATION: bindTexture: textures can not be used with multiple targets
```

Everything is fine if I also adjust the related PMREM files to use the `texture` property of their respective render targets.

tschw commented:
@vanruesc 
> Everything is fine if I also adjust the related PMREM files to use the texture property of their respective render targets.

Any clue why the behavior is not identical? Must be doing something differently... Oh, yes. Does `CubeUVRenderTarget` have a proper prototype chain, so `instanceof` will work correctly? If so, that better be fixed (as a matter of good style) and you can just change the example to pass `renderTarget.texture` as it's the new official (non-deprecated) interface.

tschw commented:
Just give it a trace in case I guessed wrong...

vanruesc committed:
Revised the previously added comments.

vanruesc committed:
Added a TODO note.

Forgot to include this one in the last commit.

tschw commented:
@vanruesc 
> However, if I modify the example so that the texture property of CubeUVRenderTarget is used directly

Call to `render` of course must keep using the render target. But I guess you mean the cases where it's actually used as a texture?

@tschw
>  Oh, yes. Does CubeUVRenderTarget have a proper prototype chain, so instanceof will work correctly?

Yes. `CubeUVRenderTarget` is actually an object constant, not a nested class. So this is **not** the problem. It's just a simple `THREE.WebGLRenderTarget`, not a cube texture, but a 2D one. If you don't get identical behavior, there's still *something* wrong :(.

vanruesc commented:
> Call to render of course must keep using the render target. But I guess you mean the cases where it's actually used as a texture?

Yes, that's what I meant.

> Any clue why the behavior is not identical? Must be doing something differently...

The modifications in this PR don't seem to be the cause of this problem.
I grabbed an unmodified version of Three and changed this line:
```js
setCubeTextureDynamic( texture.texture, slot );
```
to
```js
setCubeTextureDynamic( texture.texture || texture, slot );
```
because using `renderTarget.texture` would actually cause errors in the current state (`Uncaught TypeError: Cannot read property 'uuid' of undefined`). Then I tested the [material standard](https://github.com/mrdoob/three.js/blob/dev/examples/webgl_materials_standard.html#L203) example again and changed
```js
material.envMap = hdrCubeRenderTarget;
```
to
```js
material.envMap = hdrCubeRenderTarget.texture;
```
which resulted in the same WebGL warning and no envMap being rendered. 

I ran the example through the debugger and found out that `setTexture2D` is taking care of `hdrCubeRenderTarget`, but if its `texture` property is assigned to `material.envMap` instead, `setTextureCube` will be the place where the texture is handled. So it comes down to the preliminary detection of cube and 2D textures that determines which of the two functions is called and I still don't know where these checks happen in the library. I mean: the value of envMap must be classified somewhere, probably indirectly. 
I also think that if `hdrCubeRenderTarget.texture` was a CubeTexture right now, the example would already be broken because the simple `WebGLRenderTarget` that is bound to `envMap` wouldn't be handled by `setTexture2D` then.

I hope that makes sense.

tschw commented:
> I hope that makes sense.

Let's see if we can put it all together:


> I ran the example through the debugger and found out that setTexture2D is taking care of hdrCubeRenderTarget

The PREM stuff uses 2D textures for cube mapping and *not* a WebGL cube map. It's correct that way.

> but if its texture property is assigned to material.envMap instead, 

Ah! Yet another bubblegum and duct tape solution :). What's its type?

> setTextureCube will be the place where the texture is handled.

No, shouldn't be the case. Did you verify that it happens or is it an assumption?

In fact, it depends on how the GLSL is constructed:

> So it comes down to the preliminary detection of cube and 2D textures that determines which of the two functions is called and I still don't know where these checks happen in the library.

When there's a `sampler2D` in the GLSL `setTexture2D` is called, when there's a `samplerCube` in the GLSL `setTextureCube` is called. `WebGLUniforms` queries the compiled GLSL program.

> I mean: the value of envMap must be classified somewhere, probably indirectly.

`material.envMap.mapping` seems to be the one that decides which sampler to use in the GLSL. 

> I also think that if hdrCubeRenderTarget.texture was a CubeTexture right now, the example would already be broken because the simple WebGLRenderTarget that is bound to envMap wouldn't be handled by setTexture2D then.

No. Before uniform automation we ended up on the 2D code path because of the render target wrapper, now we end up on the 2D code path because of the sampler uniform type. Would be broken already if not; sampler mismatch is a hard error.

So the question remains: What's actually the difference when passing `.texture` instead of the render target wrapper?

vanruesc commented:
To answer your questions:

>  What's its type?

It's a simple `Texture`. The entire example setup looks perfectly fine.

> > setTextureCube will be the place where the texture is handled. 

> No, shouldn't be the case. Did you verify that it happens or is it an assumption?

I saw this behaviour in the debugger. I logged the `uuid` of `hdrCubeRenderTarget.texture` first and then observed the two texture handling functions.

> When there's a sampler2D in the GLSL setTexture2D is called, when there's a samplerCube in the GLSL setTextureCube is called. WebGLUniforms queries the compiled GLSL program.

Ok, I didn't fully understand it before, but now it's clear. Thanks :)

> material.envMap.mapping seems to be the one that decides which sampler to use in the GLSL.

This! This is the missing part I was looking for :+1: 

> So the question remains: What's actually the difference when passing .texture instead of the render target wrapper?

I've found the problem in [this line](https://github.com/mrdoob/three.js/blob/dev/examples/js/pmrem/PMREMCubeUVPacker.js#L35) of the `PMREMCubeUVPacker` class:

```js
this.CubeUVRenderTarget.mapping = THREE.CubeUVReflectionMapping;
```

I'm surprised that this doesn't yield a warning! There is no `mapping` property in [WebGLRenderTarget](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderTarget.js) anymore but there's also no [deprecation warning](https://github.com/mrdoob/three.js/blob/dev/src/Three.Legacy.js#L578). Maybe it's been removed, I don't know. Fact is, that `mapping` is dynamically added to the render target here.

Conclusion: if the `.texture` property of the render target is used, the important mapping information __is lost__. However, if the mapping info is set on the internal texture instance - as it should be - then the render target itself can't be used as the envMap. It seems that the `mapping` had to be defined on the render target in order for it to act as a 2D texture.

tschw commented:
> I'm surprised that this doesn't yield a warning! There is no mapping property in WebGLRenderTarget anymore but there's also no deprecation warning.

Maybe just a weird hack of sorts...

> However, if the mapping info is set on the internal texture instance - as it should be...

Indeed! That's how it should be.

> It seems that the mapping had to be defined on the render target in order for it to act as a 2D texture.

:+1: Seems you got it all figured out.

Just up examples & docs and this PR should be good to go!

vanruesc commented:
> Seems you got it all figured out.

Thanks to you!

> Just up examples & docs ...

I'd like to keep the example changes out of this PR - I've already prepared them for a follow-up PR. All examples will still work just fine with the changes in this PR including the PMREM demos.

I updated the WebGLRenderTarget API documentation to closely match the actual code. Just let me know if there's anything else that needs to be updated.

vanruesc committed:
Documentation: updated WebGLRenderTarget.

tschw commented:
Seems you went overachieving on the docs, a little :)

Please re-add the explanations, unless they are redundant in `Texture` (in this case there should probably be a `see [page:Texture Texture]` near the list of properties, didn't check whether it's already there).

> I updated the WebGLRenderTarget API documentation to closely match the actual code. Just let me know if there's anything else that needs to be updated.

Docs: `ShaderMaterial` - you are no longer supposed to use render targets as uniforms. Maybe `*map` properties of other materials.

> I'd like to keep the example changes out of this PR - I've already prepared them for a follow-up PR. All examples will still work just fine with the changes in this PR including the PMREM demos.

Fine. The PMREM ones are the most important ones, since they up non-trivially.

vanruesc commented:
The explanations of the texture parameters were redundant. In fact, the information about the texture `type` was outdated on the `WebGLRenderTarget` page (`HalfFloatType` was missing) and the default value for `minFilter` was incorrect (it said `LinearMipMapLinearFilter` instead of `LinearFilter`).

> there should probably be a see [page:Texture Texture] near the list of properties  

> Docs: ShaderMaterial - you are no longer supposed to use render targets as uniforms

Added these now.

> Maybe *map properties of other materials.

I'd argue against this because _every_ map (color, ao, emissive, light, alpha, normal, bump, etc.) already links to the Texture API page. I think that a warning would feel out of place.

As reference, the current explanation of the `map` property of `MeshBasicMaterial` is:
```
.map  

Set color texture map. Default is null.
```

I guess I could add something like "Must not be a WebGLRenderTarget" or "Must really be a Texture instance", but the type of the attribute should be enough. The examples are to blame for demonstrating bad practice by neglecting this.

vanruesc committed:
Docs: added more Texture info to WebGLRenderTarget.

Also removed an unnecessary html line break.

vanruesc committed:
Docs: added a notification to ShaderMaterial.

WebGLRenderTarget and WebGLRenderTargetCube must not be used as
uniforms. Their texture instance must be used instead.

tschw commented on docs/api/materials/ShaderMaterial.html:
```diff
@@ -22,6 +22,7 @@
 		<ul>
 			<li>A *ShaderMaterial* will only be rendered properly by [page:WebGLRenderer], since the GLSL code in the *vertexShader* and *fragmentShader* properties must be compiled and run on the GPU using WebGL.</li>
 			<li>As of THREE r72, directly assigning attributes in a *ShaderMaterial* is no longer supported. A [page:BufferGeometry] instance (instead of a [page:Geometry] instance) must be used instead, using [page:BufferAttribute] instances to define custom attributes.</li>
+			<li>As of THREE r76, [page:WebGLRenderTarget] or [page:WebGLRenderTargetCube] instances are no longer supposed to be used as uniforms. Their [page:Texture texture] property must be used instead.</li>
```
`r76` is already tagged. I guess the merge will probably go into `r77`.

vanruesc replied:
Whoops :sweat:

tschw commented:
@vanruesc wrote:
> It seems that the mapping had to be defined on the render target in order for it to act as a 2D texture.
> [...]
> https://github.com/vanruesc/three.js/blob/dev/docs/api/renderers/WebGLRenderTarget.html#L29-L36

I wonder whether all of the `WebGLRenderTarget` properties are still looked at.

In particular [these ones](https://github.com/vanruesc/three.js/blob/dev/docs/api/renderers/WebGLRenderTarget.html#L29-L32) as well as `anisotropy` seem to be texture properties. I guess they are copied over. Maybe just stating that in the docs would be best, just listing the names and referring to `Texture` for the description, to only document them in one place...

@tschw wrote:
> there should probably be a see [page:Texture Texture] near the list of properties 

... And it's the list I tried to refer to here. Although attention to detail and taking the docs seriously improves their quality, so note I'm not complaining about you being thorough :+1: 

> > Maybe *map properties of other materials.

@vanruesc wrote:
> I'd argue against this because every map (color, ao, emissive, light, alpha, normal, bump, etc.) already links to the Texture API page. I think that a warning would feel out of place.

Agreed. What I meant is: If there's anything that states "Texture or RenderTarget" to simply remove the "or RenderTarget" bit ;). 

> As reference, the current explanation of the map property of MeshBasicMaterial is [...]

If there's nothing like that, as seems to be the case, no update is needed, of course.

> I guess I could add something like "Must not be a WebGLRenderTarget" or "Must really be a Texture instance", but the type of the attribute should be enough.

I agree. That may in fact just cause confusion.

vanruesc commented:
> I wonder whether all of the WebGLRenderTarget properties are still looked at.  

[These](https://github.com/vanruesc/three.js/blob/dev/docs/api/renderers/WebGLRenderTarget.html#L29-L36) are only parameters. I listed them the same way as the [WebGLRenderer](https://github.com/vanruesc/three.js/blob/dev/docs/api/renderers/WebGLRenderer.html#L22-L35) does it.

The problem with `mapping` was probably just a Duck Typing phenomenon.

> In particular [these ones](https://github.com/vanruesc/three.js/blob/dev/docs/api/renderers/WebGLRenderTarget.html#L29-L32) as well as `anisotropy` seem to be texture properties. I guess they are copied over.

Yes, they are just used to construct the internal texture instance, but the important thing here is that the default value of `minFilter` is `LinearFilter` for render targets whereas it's `LinearMipMapLinearFilter` for other Textures. The `options` object also contains the `stencilBuffer` and `depthBuffer` flags. That's why I listed each item of this complex parameter explicitly with default values.

> Maybe just stating that in the docs would be best

I think that [this line](https://github.com/vanruesc/three.js/blob/dev/docs/api/renderers/WebGLRenderTarget.html#L26) alone with a plain list of the texture parameter names might not be enough for the reasons given above, if that's what you mean.

vanruesc committed:
Docs: revised a notification in ShaderMaterial.

tschw commented:
OK, looks good/ready then.

/ping @mrdoob

tschw commented:
re-
/ping @mrdoob
 
(easy to miss a ping on a release day, I figure)

tschw commented:
/ping @mrdoob

tschw commented:
/ping @mrdoob 

Anything wrong with this PR? Looks ripe and ready to me and should be worth the merge for the comments alone...

mrdoob commented:
All good. Sorry for the delay!

mrdoob commented:
Thanks guys!

mrdoob commented:
I think this PR broke `CubeCamera`...

Both `webgl_materials_cubemap_dynamic` and `webgl_materials_cubemap_dynamic2` show this:

![screen shot 2016-05-17 at 22 48 02](https://cloud.githubusercontent.com/assets/97088/15345965/7b2f440e-1c81-11e6-936f-ff9003b048bc.png)

@vanruesc do you mind taking a look?

vanruesc commented:
Hi,

The `CubeCamera` should be fine! I looked into it and dare say that this PR is unrelated to the issues at hand.

`webgl_materials_cubemap_dynamic2` _only_ shows me this:

![cubemap_dynamic2](https://cloud.githubusercontent.com/assets/9214917/15355385/95b5d682-1cf2-11e6-8acc-996eb827fa7f.png)

`webgl_materials_cubemap_dynamic` has another problem that causes the repeated WebGL warning. It shows me this in the console:

![cubemap_dynamic](https://cloud.githubusercontent.com/assets/9214917/15355425/e33d90c0-1cf2-11e6-88b8-53502f043806.png)

Quoting [DAG](http://stackoverflow.com/questions/35288245/three-js-imported-obj-model-commandbuffercontextrender-warning-render-count?answertab=votes#tab-top) from stackoverflow: "This happens when the low level render call was told to draw zero vertices/faces". So there seems to be some problem with geometry loading.

These errors/warnings in both examples have been there before this PR was merged. I went back to 16d45f5c5b7ef7ce7758dfccfc5b454627f9b903 (one commit prior to the merge) to verify this.

mrdoob commented:
Thanks! It was a bit if a wild guess. I'll bisect today.

mrdoob commented:
Never mind. Seems to be a Chrome issue, maybe 50+?

vanruesc commented:
> Seems to be a Chrome issue, maybe 50+?

I don't think that's the case. 

The warning in `webgl_materials_cubemap_dynamic2` is caused by the fact that there are three objects that use the cube map which in turn is generated by a `CubeCamera` that tries to capture the scene which contains said objects. I'm not sure why only the `Sphere` is hidden for the cube map update, but hiding all three objects solves that issue.

Current state:

```js
sphere.visible = false; // *cough*

cubeCamera.updateCubeMap( renderer, scene );

sphere.visible = true; // *cough*
```

More caughing required:

```js

cube.visible = false; // *cough*
torus.visible = false; // *cough*
sphere.visible = false; // *cough*

cubeCamera.updateCubeMap( renderer, scene );

sphere.visible = true; // *cough*
torus.visible = true; // *cough*
cube.visible = true; // *cough*
```

The errors in `webgl_materials_cubemap_dynamic` seem to be caused by something that's buried deep in the example setup. Chrome, Firefox and IE report the same repeated error at first:

```js
// inside of BufferGeometry.computeBoundingSphere
TypeError: this.attributes.position is undefined
```
 Only Chrome reports an inherent WebGL warning:
```js
[.CommandBufferContext]RENDER WARNING: Render count or primcount is 0.
```

The errors stop eventually which indicates that some geometry might simply be taking a while to load, but the setup doesn't account for that. I haven't found the offending code, though, because the example is quite large.

mrdoob commented:
Yeah. But I think that error didn't use to happen. And it's only in Chrome...

The solution is probably to have two RenderTargets and do pingponging.

I'll check with the chrome guys...
